components/sidebar/sidebar_category_list/sidebar_category_list.tsx

```diff
@@ -2,32 +2,260 @@
+            const margin = (scrollingToUnread || !this.state.showTopUnread) ? scrollMargin : scrollMarginWithUnread;
+
+            let scrollEnd;
+            const displayedChannels = this.getDisplayedChannels();
+            if (displayedChannels.length > 0 && displayedChannels[0] === channelId) {
+                // This is the first channel, so scroll right to the top
+                scrollEnd = MathUtil.mapValueInRange(0, 0, 1, 0, 1);
+            } else {
+                scrollEnd = MathUtil.mapValueInRange(top - margin, 0, 1, 0, 1);
+            }
+
+            this.scrollToPosition(scrollEnd);
+        } else if (bottom > scrollTop + scrollHeight) {
+            // Scroll down to the item
+            const margin = (scrollingToUnread || !this.state.showBottomUnread) ? scrollMargin : scrollMarginWithUnread;
+            const scrollEnd = (bottom - scrollHeight) + margin;
+
+            this.scrollToPosition(scrollEnd);
+        }
+    }
+
+    scrollToPosition = (scrollEnd: number) => {
+        // Stop the current animation before scrolling
+        this.scrollAnimation.setCurrentValue(this.scrollbar.current!.scrollTop).setAtRest();
+
+        this.scrollAnimation.setEndValue(scrollEnd);
+    }
+
+    updateUnreadIndicators = () => {
+        let showTopUnread = false;
+        let showBottomUnread = false;
+
+        // Consider partially obscured channels as above/below
+        const firstUnreadChannel = this.getFirstUnreadChannel();
+        const lastUnreadChannel = this.getLastUnreadChannel();
+
+        if (firstUnreadChannel) {
+            const firstUnreadElement = this.channelRefs.get(firstUnreadChannel);
+            const firstUnreadPosition = firstUnreadElement ? firstUnreadElement.offsetTop : null;
+
+            if (firstUnreadPosition && ((firstUnreadPosition + firstUnreadElement!.offsetHeight) - scrollMargin) < this.scrollbar.current!.scrollTop) {
+                showTopUnread = true;
+            }
+        }
+
+        if (lastUnreadChannel) {
+            const lastUnreadElement = this.channelRefs.get(lastUnreadChannel);
+            const lastUnreadPosition = lastUnreadElement ? lastUnreadElement.offsetTop : null;
+
+            if (lastUnreadPosition && (lastUnreadPosition + scrollMargin) > (this.scrollbar.current!.scrollTop + this.scrollbar.current!.clientHeight)) {
+                showBottomUnread = true;
+            }
+        }
+        if (showTopUnread !== this.state.showTopUnread || showBottomUnread !== this.state.showBottomUnread) {
+            this.setState({
+                showTopUnread,
+                showBottomUnread,
+            });
+        }
+    }
+
+    getFirstUnreadChannel = () => {
+        return this.getFirstUnreadChannelFromChannelIdArray(this.getDisplayedChannels());
+    }
+
+    getLastUnreadChannel = () => {
+        return this.getFirstUnreadChannelFromChannelIdArray(this.getDisplayedChannels().reverse());
+    }
+
+    getDisplayedChannels = () => {
```

Comment:
Nothing to change for now, but I think this component might not be able to access a list of the displayed channels at some point. We might need to think about how that could work in the future. Maybe something to do with `setChannelRef` where it tells this component where the channel is in the sidebar and whether or not its unread?